src/node_snapshotable.cc

```diff
@@ -1111,12 +1100,14 @@ void DeserializeNodeInternalFields(Local<Object> holder,
                        (*holder),                                              \
                        NativeTypeName::type_name.c_str());                     \
     env_ptr->EnqueueDeserializeRequest(                                        \
-        NativeTypeName::Deserialize, holder, index, info->Copy());             \
+        NativeTypeName::Deserialize,                                           \
+        holder,                                                                \
+        index,                                                                 \
+        info->Copy<NativeTypeName::InternalFieldInfo>());                      \
     break;                                                                     \
   }
     SERIALIZABLE_OBJECT_TYPES(V)
 #undef V
-    default: { UNREACHABLE(); }
```

Comment:
Can we warn here about the unrecognizable type of the embedder object instead? It seems to me that the conditions can be like switching between the top of the trunk of development with the same snapshot metadata introduced in https://github.com/nodejs/node/pull/44132. Ignoring it with no message is not helpful to identify what's going wrong.

Reply:
Programmatically it  should not be reachable here though - if you simply add a new type to SERIALIZABLE_OBJECT_TYPES without implementing the serialization methods for the Snapshotable class, it should not compile. If you make any BaseObject inherit from Snapshotable without adding a type for it, it does not compile either. BaseObjects that are does not inherit from Snapshotable are ignored during serialization either way and it'll crash due to unregistered global handles (you can usually guess from the object information printed by the crash what kind of object it is e.g. with WeakReferences holding on to a diagnostics channel, you will see a Channel)

Reply:
Yeah, it is true if the external snapshot data is used for the same node binary.

However, sharing the external snapshot data accidentally between two same-metadata Node.js versions, like in development, this branch can be still reachable. When the relevant JS object is being accessed later, the program simply crashes as the native object is not correctly deserialized. A simple reproduction can be compiling https://github.com/nodejs/node/pull/44193 and building a snapshot with a WeakReference then resurrecting node with  the binary compiled with this PR:

```
// Build snapshot with https://github.com/nodejs/node/pull/44193
const { internalBinding } = require('internal/test/binding');
const { WeakReference } = internalBinding('util');
const v8 = require('v8');
globalThis.foo = new WeakReference({});
console.log(globalThis.foo.get());

v8.startupSnapshot.setDeserializeMainFunction(() => {
  // Resurrect snapshot with this PR. Here it crashes.
  console.log(globalThis.foo.get());
});
```

Reply:
Thanks, added some code to print more about the unknown embedder object types.